### PR TITLE
i#1312 AVX-512 support: Add AVX-512 VEX encoded scalar opmask instructions.

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds the AVX-512 VEX encoded scalar opmask instructions and extends the existing VEX
tables. Unfortunately there are no Appendix A/B style documented Intel manuals available
for AVX-512 in the same format as the existing x86 instruction sets. Therefore, the introduced
type name 'K' is invented by us, based on the Intel AVX-512 instruction descriptions.

Makes necessary modification in decoder to support VEX encoded AVX-512 instructions,
specifically decoding and encoding more cases of the vex W opcode extension bit.

Adds support to encoder for new VEX encodings. Fixes a bug to encode primary opcode byte
in 3-byte VEX versions. Adds support for prefix_ext indirection to after vex_ext in decoder.

Adds new operand types TYPE_K_REG, TYPE_K_MODRM, TYPE_K_MODRM_R, TYPE_K_VEX, TYPE_K_EVEX.

Adds all INSTR_CREATE macros for the new instructions.

Adds all test macros to x86 api.ir test.

Adds raw encodings decoder test to x86 common.decode test.

Fixes existing disassemble test.

Issue: #1312

### CODE REVIEW

File: core/arch/x86/opcode.h [hunk 1]
```diff
@@ -1266,6 +1266,59 @@ enum {
     /* 1105 */ OP_adox, /**< IA-32/AMD64 adox opcode. */
     /* 1106 */ OP_adcx, /**< IA-32/AMD64 adox opcode. */
 
+    /* Intel AVX-512 */
+    /* 1107 */ OP_kmovw,    /**< IA-32/AMD64 AVX-512 kmovw opcode */
```
Comment (derekbruening) [inline core/arch/x86/opcode.h]: All the existing doxygen comments in a period.  Best to match them for consistency, to match the style guide, etc.

File: core/arch/aarch64/instr.c [hunk 2]
```diff
@@ -136,6 +136,12 @@ instr_is_return(instr_t *instr)
     return (opc == OP_ret);
```
Comment (derekbruening) [inline core/arch/aarch64/instr.c]: > Unfortunately there are no Appendix A/B style documented Intel manuals available for AVX-512 in the same format as the existing x86 instruction sets.

Strange.  Do the AMD manuals have them in there?

> Add support to encoder for new VEX encodings. Fixed a bug to encode primary opcode byte in 3-byte VEX versions.

grammar consistency: s/Add/Adds/; s/Fixed/Fixes/.

> Adds raw encodings decoder test to x86 common.decoder test.

It's common.decode, not common.decoder.

File: core/arch/x86/opcode.h (hunk 1 above)
Comment (hgreving2304) [reply core/arch/x86/opcode.h]: Done (will update).

File: core/arch/aarch64/instr.c (hunk 2 above)
Comment (hgreving2304) [reply core/arch/aarch64/instr.c]: > Strange. Do the AMD manuals have them in there?

Don't think AMD even support AVX-512. No manuals yet.

> grammar consistency: s/Add/Adds/; s/Fixed/Fixes/.

Fixed.

> It's common.decode, not common.decoder.

Fixed.

File: core/arch/x86/decode.c [hunk 3]
```diff
@@ -1828,6 +1837,40 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+    case TYPE_K_VEX: {
+        /* part of AVX-512: vex.vvvv selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_VEX, di, optype, opsize));
+        /* AVX-512 VEX encoded scalar opmask instructions, the size is determined
+         * by the opcode.
+         */
+        return true;
+    }
+    case TYPE_K_EVEX: {
+        CLIENT_ASSERT(false, "decode error: unsupported.");
```
Comment (derekbruening) [inline core/arch/x86/decode.c]: As you know an assert like this should not be used for a legitimate code path.  If this is not yet implemented as part of incremental code it needs clear labeling as such in a comment and the output message.

File: core/arch/x86/decode.c [hunk 4]
```diff
@@ -709,7 +709,7 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
 }
 
 /* Given an instr_info_t PREFIX_EXT entry, reads the next entry based on the prefixes.
- * Note that this function does not initialise the opcode field in \p di but is set in
+ * Note that this function does not initialize the opcode field in \p di but is set in
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode.c]: I can only support countries that are part of the European union, that excludes Britain :) No, I thought it somehow stood out since it's spelled nowhere else like that. A quick grep does find it in third_party/valgrind.

File: core/arch/x86/decode.c [hunk 5]
```diff
@@ -1828,6 +1837,40 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
         /* no need to set size as it's a GPR */
         return true;
     }
+    case TYPE_K_MODRM: {
+        /* part of AVX-512: modr/m.rm selects opmask register or mem addr */
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Fixed.

File: core/arch/x86/decode.c [hunk 6]
```diff
@@ -1828,6 +1837,40 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_RM, di, optype, opsize));
+        /* For VEX encoded scalar opmask instructions, the size is determined
+         * by the opcode.
+         */
+        return true;
+    }
+    case TYPE_K_REG: {
+        /* part of AVX-512: modr/m.reg selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_REG, di, optype, opsize));
+        /* AVX-512 VEX encoded scalar opmask instructions, the size is determined
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Fixed.

File: core/arch/x86/decode.c [hunk 7]
```diff
@@ -1828,6 +1837,40 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+         * by the opcode.
+         */
+        return true;
+    }
+    case TYPE_K_VEX: {
+        /* part of AVX-512: vex.vvvv selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_VEX, di, optype, opsize));
+        /* AVX-512 VEX encoded scalar opmask instructions, the size is determined
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Fixed.

File: core/arch/x86/decode.c (hunk 3 above)
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Done.

File: core/arch/x86/encode.c [hunk 8]
```diff
@@ -844,6 +853,12 @@ reg_size_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, reg_
         }
         return true;
     }
+    if (optype == TYPE_K_REG || optype == TYPE_K_MODRM || optype == TYPE_K_MODRM_R ||
+        optype == TYPE_K_VEX) {
+        return (opsize == OPSZ_1 || opsize == OPSZ_2 || opsize == OPSZ_4 ||
+                opsize == OPSZ_8);
+    }
+    /* XXX i#1312: TYPE_K_EVEX will be of variable size. */
```
Comment (derekbruening) [inline core/arch/x86/encode.c]: "will be": but I thought this PR was adding support for that type?  What makes the existing size_ok() check not work there?

File: core/arch/x86/encode.c [hunk 9]
```diff
@@ -844,6 +853,12 @@ reg_size_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, reg_
         }
         return true;
     }
+    if (optype == TYPE_K_REG || optype == TYPE_K_MODRM || optype == TYPE_K_MODRM_R ||
+        optype == TYPE_K_VEX) {
+        return (opsize == OPSZ_1 || opsize == OPSZ_2 || opsize == OPSZ_4 ||
+                opsize == OPSZ_8);
```
Comment (derekbruening) [inline core/arch/x86/encode.c]: Won't the size_ok() call right above handle these?  What is the reason they need special code?

File: core/arch/x86/encode.c [hunk 10]
```diff
@@ -1788,9 +1825,11 @@ encode_operand(decode_info_t *di, int optype, opnd_size_t opsize, opnd_t opnd)
     case TYPE_E:
     case TYPE_Q:
     case TYPE_W:
-    case TYPE_R:       /* we already ensured this is a reg, not memory */
-    case TYPE_P_MODRM: /* we already ensured this is a reg, not memory */
-    case TYPE_V_MODRM: /* we already ensured this is a reg, not memory */
+    case TYPE_K_MODRM:
```
Comment (derekbruening) [inline core/arch/x86/encode.c]: nit: Just seems an unusual ordering, K inserted up here, but K R appended?

File: core/arch/x86/encode.c [hunk 11]
```diff
@@ -2589,10 +2643,11 @@ instr_encode_arch(dcontext_t *dcontext, instr_t *instr, byte *copy_pc, byte *fin
     }
 
     /* second opcode byte, if there is one */
-    if (TEST(OPCODE_TWOBYTES, info->opcode)) {
+    if (TEST(OPCODE_TWOBYTES, info->opcode) || TEST(OPCODE_THREEBYTES, info->opcode)) {
```
Comment (derekbruening) [inline core/arch/x86/encode.c]: This does not look avx-512 specific -- are you sure this is needed?  Wouldn't this mean sthg was broken before?

File: core/arch/x86/encode.c [hunk 12]
```diff
@@ -2479,7 +2529,8 @@ instr_encode_arch(dcontext_t *dcontext, instr_t *instr, byte *copy_pc, byte *fin
         encode_operand(&di, info->src2_type, info->src2_size, instr_get_src(instr, 1));
     if (info->src3_type != TYPE_NONE)
         encode_operand(&di, info->src3_type, info->src3_size, instr_get_src(instr, 2));
-    if (di.mod == 5 && di.reg < 8) { /* mod may never be set (e.g., OP_extrq) */
+    if (!opc_is_opmask_arch(di.opcode) && di.mod == 5 &&
```
Comment (derekbruening) [inline core/arch/x86/encode.c]: I think we'd rather use a bit or flag inside info rather than checking a list of opcodes.

File: core/arch/x86/instr_create.h [hunk 13]
```diff
@@ -1396,6 +1396,27 @@
     instr_create_1dst_1src((dc), OP_vphsubwd, (d), (s))
 #define INSTR_CREATE_vphsubdq(dc, d, s) \
     instr_create_1dst_1src((dc), OP_vphsubdq, (d), (s))
+/* AVX-512 VEX */
```
Comment (derekbruening) [inline core/arch/x86/instr_create.h]: nit: Better to append instead of inserting here in between AVX and AVX2.

File: suite/tests/api/ir_x86.c [hunk 14]
```diff
@@ -105,6 +105,7 @@ static byte buf[8192];
 #define IMMARG(sz) opnd_create_immed_int(37, sz)
 #define TGTARG opnd_create_instr(instrlist_last(ilist))
 #define REGARG(reg) opnd_create_reg(REG_##reg)
+#define DR_REGARG(reg) opnd_create_reg(DR_REG_##reg)
```
Comment (derekbruening) [inline suite/tests/api/ir_x86.c]: I would say just change REGARG: no need for a separate thing.

File: core/arch/x86/decode.c [hunk 15]
```diff
@@ -1828,6 +1837,41 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_RM, di, optype, opsize));
+        /* For VEX encoded scalar opmask instructions, the size is determined
+         * by the opcode.
+         */
+        return true;
+    }
+    case TYPE_K_REG: {
+        /* part of AVX-512: modrm.reg selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_REG, di, optype, opsize));
+        /* for AVX-512 VEX encoded scalar opmask instructions, the size is determined
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Fixed.

File: core/arch/x86/decode_private.h [hunk 16]
```diff
@@ -252,6 +252,14 @@ enum {
...
 /* Predicated via something complex */
 #define HAS_PRED_COMPLEX 0x0800
+/* Instr must be encoded with VEX.W=1.  If VEX.W=0 this is an invalid instr.
+ * While we still have vex_W_extensions entries, the encoder needs this flag.
+ * That's because for some VEX instructions this is an opcode extension. If
+ * this bit is badly needed (e.g. if we're running out of space in the flags
+ * field, then this also could get encoded as part of the opcode as a new
+ * leading byte).
+ */
+#define REQUIRES_VEX_W_1 0x1000
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode_private.h]: Great, that works (encoder has to test the bit as well). Done.

File: core/arch/x86/decode.c [hunk 17]
```diff
@@ -1828,6 +1837,41 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+    case TYPE_K_VEX: {
+        /* part of AVX-512: vex.vvvv selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_VEX, di, optype, opsize));
+        /* For AVX-512 VEX encoded scalar opmask instructions, the size is determined
+         * by the opcode.
+         */
+        return true;
+    }
+    case TYPE_K_EVEX: {
+        /* XXX i#1312: will be supported as part of the AVX-512 EVEX encodings. */
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Done (TODO).

File: core/arch/x86/encode.c (hunk 8 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: That's a comment for TYPE_K_EVEX. This type is not used. The instructions in this patch are just the VEX versions. The much larger set of EVEX instructions uses the mask as an actual mask, and the register is encoded in one of the EVEX bytes (evex.aaa). This is the type to be used. It's not 100% clear to me yet whether we need a variable type or if a fixed type will be ok. First I thought we need a variable type but now I think we will have 4 fixed size types again, because I think the size should always be defined by the opcode.

File: core/arch/x86/encode.c (hunk 9 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: The existing code make its decision based on the operand and prefixes. The special thing about these scalar k register instructions is that the k register size is implied by its opcode (w, b, q, d versions). One could move this into the size_ok function, but that would simply be a re-factor. I think outside makes more sense.

File: core/arch/x86/encode.c (hunk 10 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: Ah ok yes I did that intentionally. The reason is that I put the "we already ensured this is a reg, not memory" one in a group. It's important to know that these ones are always (verified already) a register. The TYPE_K_MODRM can also be memory, so I pulled it up before the group starts. If that doesn't make sense please re-open.

File: core/arch/x86/encode.c (hunk 11 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: That was a bug IMO. But apparently no instructions needed it. The encoder encode the primary opcode for three byte versions if that's missing. I think there may not have been any three-byte VEX opcodes until now. I didn't see any when I checked.

File: suite/tests/api/ir_x86.c (hunk 14 above)
Comment (hgreving2304) [reply suite/tests/api/ir_x86.c]: Done.

File: core/arch/x86/instr_create.h (hunk 13 above)
Comment (hgreving2304) [reply core/arch/x86/instr_create.h]: Done.

File: core/arch/x86/encode.c (hunk 12 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: There are a few uses alike like instr_is_cti, instr_get_predicate etc. Before I do change it, how exactly? You mean flag in the decoder table (nah..)? Or do you mean just a bool flag like bool is_opmask = opc_is_opmask_arch(di.opcode) further up? If not a flag in the table, which I think is unnecessary, then one has to check the opcode _somewhere_.

File: core/arch/x86/decode_table.c [hunk 18]
```diff
@@ -4299,8 +4376,147 @@ const instr_info_t prefix_extensions[][8] = {
     {INVALID,      0x6638f618, "(bad)",   xx, xx, xx, xx, xx, no, x, NA},
     {OP_mulx,      0xf238f618, "mulx",    By, Gy, Ey, uDX, xx, mrm|vex, x, END_LIST},
   },
+  { /* prefix extension 144 */
+    {OP_seto,        0x0f9010, "seto",    Eb, xx, xx, xx, xx, mrm, fRO, END_LIST},
+    {INVALID,      0xf30f9010, "(bad)",   xx, xx, xx, xx, xx, no, x, NA},
+    {INVALID,      0x660f9010, "(bad)",   xx, xx, xx, xx, xx, no, x, NA},
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: I don't think this is invalid: this is just "data16 seto" where the data16 is ignored.  I just tried it and it ran fine on my processor; plus multiple decoders accept it just fine:
```
$ echo " 0x66 0x0f 0x91 0x7d 0xfc" | /usr/bin/llvm-mc -arch x86 --disassemble
	.text
	setno	-4(%ebp)
$ echo " 0x66 0x0f 0x91 0x7d 0xfc" | /extsw/pkgs/disasm/capstone/build/capstone -x86 - 
+0x0000  660f917dfc   setno	byte ptr [ebp - 4]
(gdb) set {unsigned char[400]}0x04311000 = { 66, f, 90, 7d, fc, 90, 90, 90, 90, 90, 90 }
(gdb) x/3i 0x04311000
   0x4311000:	data16 seto -0x4(%rbp)
   0x4311005:	nop
```

I think this needs to instead split by vex encoding to distinguish seto from these new instructions, instead of splitting by prefix first.

File: core/arch/x86/decode_table.c [hunk 19]
```diff
@@ -1279,6 +1279,63 @@ const instr_info_t * const op_instr[] =
     /* OP_jmpe      */   &base_extensions[13][6],
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/arch/x86/decode_table.c]: Filed #3563

File: core/arch/x86/encode.c (hunk 12 above)
Comment (derekbruening) [reply core/arch/x86/encode.c]: A flag in the table.  Why have an explicit series of branches looking at opcodes when we have per-opcode info right there in the table entry?

File: core/arch/x86/encode.c (hunk 9 above)
Comment (derekbruening) [reply core/arch/x86/encode.c]: There must be precedent for this: does it really need special handling differently from other registers?  Various SIMD opcodes operate on different parts of the *mm registers as specified by the opcode and we don't have special code here for them.  Maybe these mask registers should be treated like the sub-pieces of the *mm registers.  (The other way is to give the sub-registers separate names: like al, ax, eax, rax.)

File: core/arch/x86/encode.c (hunk 12 above)
Comment (derekbruening) [reply core/arch/x86/encode.c]: > There are a few uses alike like instr_is_cti, instr_get_predicate etc.

During encoding?  I wouldn't think so: the encoder should not go and query things externally based on the opcode.  The encoding table entry should have everything it needs.  It should not be dispatching on opcode: that's the whole point of the encoding table, to have everything in one spot.

File: core/arch/x86/encode.c (hunk 11 above)
Comment (derekbruening) [reply core/arch/x86/encode.c]: It looks like OPCODE_TWOBYTES is always set when OPCODE_THREEBYTES is set so I do not think there is any bug here.  Instead, the decode_private.h docs should make that explicit.

File: core/arch/x86/decode_table.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Moved. Done.

File: core/arch/x86/decode_table.c [hunk 20]
```diff
@@ -1412,6 +1469,27 @@ const instr_info_t * const op_instr[] =
 #define Lss TYPE_L, OPSZ_4_of_16 /* immed is 1 byte but reg is xmm/ymm */
 #define Lsd TYPE_L, OPSZ_8_of_16 /* immed is 1 byte but reg is xmm/ymm */
 
+/* AVX-512 additions */
+#define Krb TYPE_K_REG, OPSZ_1
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Done.

[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: The K type fits very well with the k registers. I've renamed the float K* into F*.

[the remark this answers is not in the document]
Comment (derekbruening) [reply core/arch/x86/decode_table.c]: Is that going to be confusing vs Intel's type F == "EFLAGS/RFLAGS Register"?  At least worth tweaking the existing "I don't use type F" comment and one by the F defines.

File: core/arch/x86/decode_table.c (hunk 18 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: If that's the case I can go through VEX_EXT first. Done. The decoder now checks for PREFIX_EXT after VEXT_EXT, which is needed for the new instructions.

File: core/arch/x86/encode.c (hunk 12 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: Good that I (had to) look at this again. I had fixed a bug (in my patch) to support TYPE_K_MODRM k registers OR memory. The memory part through encoding modrm was missing. Now modrm.mod will always be set by the operands, the special opmask case here is not needed anymore. I removed it, including the functions I had added. Should be clean now.

**Comment by hgreving2304** [comment]
PTAL

**Comment by hgreving2304** [comment]
(PTAL)

**Comment by hgreving2304** [comment]
PTAL.

**Comment by hgreving2304** [comment]
Merged with IA32_ON_IA64 upstream (resolved conflicts).

File: core/arch/x86/decode.c [hunk 21]
```diff
@@ -1809,6 +1824,41 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+        if (di->mod != 3) {
+            return decode_modrm(di, optype, opsize, NULL, opnd);
+        }
+        /* fall through*/
+    }
+    case TYPE_K_MODRM_R: {
+        /* part of AVX-512: modrm.rm selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_RM, di, optype, opsize));
+        /* For VEX encoded scalar opmask instructions, the size is determined
+         * by the opcode.
```
Comment (derekbruening) [inline core/arch/x86/decode.c]: I don't understand this comment: what is it telling the reader?  We have the opcode-determined size here in the `opsize` variable.  Should we call opnd_set_size(opnd, opsize)?  Do we need this comment?  Ditto below.

File: core/arch/x86/decode.c [hunk 22]
```diff
@@ -1809,6 +1824,41 @@ decode_operand(decode_info_t *di, byte optype, opnd_size_t opsize, opnd_t *opnd)
...
+        /* part of AVX-512: vex.vvvv selects opmask register */
+        *opnd = opnd_create_reg(decode_reg(DECODE_REG_VEX, di, optype, opsize));
+        /* For AVX-512 VEX encoded scalar opmask instructions, the size is determined
+         * by the opcode.
+         */
+        return true;
+    }
+    case TYPE_K_EVEX: {
+        /* TODO i#1312: will be supported as part of the AVX-512 EVEX encodings. */
+        CLIENT_ASSERT(false, "XXX i#1312: decode error: unsupported yet.");
```
Comment (derekbruening) [inline core/arch/x86/decode.c]: FIXME not XXX

File: core/arch/x86/decode_table.c [hunk 23]
```diff
@@ -1273,6 +1273,62 @@ const instr_info_t * const op_instr[] =
...
+    /* OP_kshiftrb      */  &vex_W_extensions[102][0],
+    /* OP_kshiftrq      */  &vex_W_extensions[103][1],
+    /* OP_kshiftrd      */  &vex_W_extensions[103][0],
+    /* OP_ktestw        */  &vex_W_extensions[104][0],
+    /* OP_ktestb        */  &vex_W_extensions[105][0],
+    /* OP_ktestq        */  &vex_W_extensions[104][1],
+    /* OP_ktestd        */  &vex_W_extensions[105][1],
+
+    /* AVX-512 EVEX encoded */
+    /* XXX i#1312: TODO. */
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: XXX is an optional change or improvement which this is not: should be FIXME or if we add TODO to the style guide (sounds fine to me: feel free to edit the wiki page to add it) then TODO.

File: core/arch/x86/decode_table.c [hunk 24]
```diff
@@ -1430,12 +1508,12 @@ const instr_info_t * const op_instr[] =
 #define Zb  TYPE_XLAT, OPSZ_1
 #define Bq  TYPE_MASKMOVQ, OPSZ_8
 #define Bdq  TYPE_MASKMOVQ, OPSZ_16
-#define Kw  TYPE_FLOATMEM, OPSZ_2
-#define Kd  TYPE_FLOATMEM, OPSZ_4
-#define Kq  TYPE_FLOATMEM, OPSZ_8
-#define Kx  TYPE_FLOATMEM, OPSZ_10
-#define Ky  TYPE_FLOATMEM, OPSZ_28_short14 /* _14_ if data16 */
-#define Kz  TYPE_FLOATMEM, OPSZ_108_short94 /* _98_ if data16 */
+#define Fw  TYPE_FLOATMEM, OPSZ_2
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: I think this rename should be a separate PR.

File: core/arch/x86/decode_table.c [hunk 25]
```diff
@@ -1406,6 +1462,28 @@ const instr_info_t * const op_instr[] =
...
+/* AVX-512 additions */
+#define KRb TYPE_K_REG, OPSZ_1
+#define KRw TYPE_K_REG, OPSZ_2
+#define KRd TYPE_K_REG, OPSZ_4
+#define KRq TYPE_K_REG, OPSZ_8
+#define KRMb TYPE_K_MODRM_R, OPSZ_1
+#define KRMw TYPE_K_MODRM_R, OPSZ_2
+#define KRMd TYPE_K_MODRM_R, OPSZ_4
+#define KRMq TYPE_K_MODRM_R, OPSZ_8
+#define KMRMb TYPE_K_MODRM, OPSZ_1
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: How about modeling after types P, Q, and R?  So s/KR/KP/; s/KRM/KR/; s/KMRM/KQ/.  It's then closer to existing names and the table stays compact.  WDYT?

File: core/arch/x86/decode_table.c [hunk 26]
```diff
@@ -4522,6 +4738,48 @@ const instr_info_t vex_extensions[][2] = {
...
     {INVALID,   0x66380d18, "(bad)", xx, xx, xx, xx, xx, no, x, NA},
     {OP_vpermilpd, 0x66380d18, "vpermilpd", Vx, xx, Hx,Wx, xx, mrm|vex|reqp, x, END_LIST},
+  }, { /* vex ext 79 */
+    {OP_seto,    0x0f9010,             "seto", Eb, xx, xx, xx, xx, mrm, fRO, END_LIST},
+    {PREFIX_EXT, 0x0f9010, "(prefix ext 144)", xx, xx, xx, xx, xx, mrm,   x, 144},
+  }, { /* vex ext 80 */
+    {OP_setno,   0x0f9110,            "setno", Eb, xx, xx, xx, xx, mrm, fRO, END_LIST},
+    {PREFIX_EXT, 0x0f9110, "(prefix ext 145)", xx, xx, xx, xx, xx, mrm,   x, 145},
+  }, { /* vex ext 81 */
+    {OP_setb,    0x0f9210,             "setb", Eb, xx, xx, xx, xx, mrm, fRO, END_LIST},
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: bug: setb is fRC!  Please double-check all the others.  Are there more pre-review tests that could be run (and prob mentioned in the commit message) to give the reviewer confidence these are correct?

File: core/arch/x86/encode.c (hunk 9 above)
Comment (derekbruening) [reply core/arch/x86/encode.c]: This still doesn't seem necessary: size_ok collapses sub-reg sizes.  Are the smaller sizes sub-reg sizes, just like pieces of xmm registers?  It seems like they are, and thus size_ok will work fine, if we make them OPSZ_{1,2,4}_of_8.

File: suite/tests/api/ir_x86_2args_avx512_vex.h [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+OPCODE(kmovb_km, kmovb, kmovb, 0, REGARG(K1), MEMARG(OPSZ_1))
+OPCODE(kmovq_km, kmovq, kmovq, 0, REGARG(K2), MEMARG(OPSZ_8))
+OPCODE(kmovd_km, kmovd, kmovd, 0, REGARG(K3), MEMARG(OPSZ_4))
+OPCODE(kmovw_mk, kmovw, kmovw, 0, MEMARG(OPSZ_2), REGARG(K4))
+OPCODE(kmovb_mk, kmovb, kmovb, 0, MEMARG(OPSZ_1), REGARG(K5))
+OPCODE(kmovq_mk, kmovq, kmovq, 0, MEMARG(OPSZ_8), REGARG(K6))
+OPCODE(kmovd_mk, kmovd, kmovd, 0, MEMARG(OPSZ_4), REGARG(K7))
+OPCODE(kmovw_kr, kmovw, kmovw, 0, REGARG(K0), REGARG(EAX))
+OPCODE(kmovb_kr, kmovb, kmovb, 0, REGARG(K1), REGARG(EBX))
+#ifdef X64
```
Comment (derekbruening) [inline suite/tests/api/ir_x86_2args_avx512_vex.h]: I don't think we use ifdefs in any of these files: use the x64-only param like the existing code does.  Ditto below.

File: suite/tests/common/decode.c [hunk 28]
```diff
@@ -366,6 +376,82 @@ GLOBAL_LABEL(FUNCNAME:)
...
+        RAW(c4) RAW(e1) RAW(f9) RAW(99) RAW(fe)         /* ktestd %k6,%k7 */
+        RAW(c4) RAW(e3) RAW(f9) RAW(32) RAW(c8) RAW(ff) /* kshiftlw $0xff,%k0,%k1 */
+        RAW(c4) RAW(e3) RAW(79) RAW(32) RAW(da) RAW(7b) /* kshiftlb $0x7b,%k2,%k3 */
+        RAW(c4) RAW(e3) RAW(f9) RAW(33) RAW(ec) RAW(07) /* kshiftlq $0x7,%k4,%k5 */
+        RAW(c4) RAW(e3) RAW(79) RAW(33) RAW(fe) RAW(63) /* kshiftld $0x63,%k6,%k7 */
+        RAW(c4) RAW(e3) RAW(f9) RAW(30) RAW(c8) RAW(df) /* kshiftrw $0xdf,%k0,%k1 */
+        RAW(c4) RAW(e3) RAW(79) RAW(30) RAW(da) RAW(65) /* kshiftrb $0x65,%k2,%k3 */
+        RAW(c4) RAW(e3) RAW(f9) RAW(31) RAW(ec) RAW(05) /* kshiftrq $0x5,%k4,%k5 */
+        RAW(c4) RAW(e3) RAW(79) RAW(31) RAW(fe) RAW(2f) /* kshiftrd $0x2f,%k6,%k7 */
+        RAW(c4) RAW(e1) RAW(79) RAW(31) RAW(fe) RAW(2f) /* bad */
```
Comment (derekbruening) [inline suite/tests/common/decode.c]: How does this test tell us whether it properly decoded any of the above?  It seems like it should have two separate functions, one with nothing bad which does not fault (has to check for avx512 support on processor first) and one with sthg bad.

File: core/arch/x86/decode.c (hunk 21 above)
Comment (hgreving2304) [reply core/arch/x86/decode.c]: I thought it helps to understand that the opmask instruction defines the k register's size implicitly by its opcode. Removed the comment(s), done.

File: core/arch/x86/decode.c (hunk 22 above)
Comment (hgreving2304) [reply core/arch/x86/decode.c]: I think we came up with "TODO". Done.

File: core/arch/x86/decode_table.c (hunk 23 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Done.

File: core/arch/x86/decode_table.c (hunk 26 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Fixed. I've checked the others manually.

File: core/arch/x86/decode_table.c (hunk 24 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Done: https://github.com/DynamoRIO/dynamorio/pull/3572

File: suite/tests/api/ir_x86_2args_avx512_vex.h (hunk 27 above)
Comment (hgreving2304) [reply suite/tests/api/ir_x86_2args_avx512_vex.h]: Done.

File: core/arch/x86/decode_table.c (hunk 25 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Ok, fine with that. Done.

File: suite/tests/common/decode.c (hunk 28 above)
Comment (hgreving2304) [reply suite/tests/common/decode.c]: Note that I put this in as a bonus, AVX(2) didn't have anything. I took this from the sse3 and 3dnow test, they seem to have the same issue. But you're right, the "bad" instruction doesn't make sense. I've removed the "bad" instance from the test.

**Comment by hgreving2304** [comment]
> Please look it over carefully and do some testing vs external decoders. The bugs found during review make me worried that there are others...

I am looking into the gas testsuite and whether we can use it. I wonder if it's possible to extract the disassembler part.
